src/transport/SecureSessionTable.cpp

```diff
@@ -0,0 +1,277 @@
+            if (false)
+            {
+                return Optional<SessionHandle>::Missing();
+            }
+
+            (void) fabricIndex;
+        }
+    }
+
+    SecureSession * result = mEntries.CreateObject(*this, secureSessionType, localSessionId, localNodeId, peerNodeId, peerCATs,
+                                                   peerSessionId, fabricIndex, config);
+    return result != nullptr ? MakeOptional<SessionHandle>(*result) : Optional<SessionHandle>::Missing();
+}
+
+Optional<SessionHandle> SecureSessionTable::CreateNewSecureSession(SecureSession::Type secureSessionType,
+                                                                   ScopedNodeId sessionEvictionHint)
+{
+    Optional<SessionHandle> rv = Optional<SessionHandle>::Missing();
+    SecureSession * allocated  = nullptr;
+
+    auto sessionId = FindUnusedSessionId();
+    VerifyOrExit(sessionId.HasValue(), rv = Optional<SessionHandle>::Missing());
+
+    //
+    // We allocate a new session out of the pool if we have space in it. If we don't, we need
+    // to run the eviction algorithm to get a free slot. We shall ALWAYS be guaranteed to evict
+    // an existing session in the table in normal operating circumstances.
+    //
+    if (mEntries.Allocated() < CHIP_CONFIG_SECURE_SESSION_POOL_SIZE)
+    {
+        allocated = mEntries.CreateObject(*this, secureSessionType, sessionId.Value());
+        VerifyOrExit(allocated != nullptr, rv = Optional<SessionHandle>::Missing());
+    }
+    else
+    {
+        allocated = EvictAndAllocate(sessionId.Value(), secureSessionType, sessionEvictionHint);
+    }
+
+    rv             = MakeOptional<SessionHandle>(*allocated);
+    mNextSessionId = sessionId.Value() == kMaxSessionID ? static_cast<uint16_t>(kUnsecuredSessionId + 1)
+                                                        : static_cast<uint16_t>(sessionId.Value() + 1);
+
+exit:
+    return rv;
+}
+
+SecureSession * SecureSessionTable::EvictAndAllocate(uint16_t localSessionId, SecureSession::Type secureSessionType,
+                                                     const ScopedNodeId & sessionEvictionHint)
+{
+    VerifyOrDieWithMsg(!mRunningEvictionLogic, SecureChannel,
+                       "EvictAndAllocate isn't re-entrant, yet someone called us while we're already running");
+    mRunningEvictionLogic = true;
+
+    ChipLogProgress(SecureChannel, "Evicting a slot for session with LSID: %d, type: %u", localSessionId,
+                    (uint8_t) secureSessionType);
+
+    VerifyOrDie(mEntries.Allocated() <= CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
+
+    //
+    // Create a temporary list of objects each of which points to a session in the existing
+    // session table, but are swappable. This allows them to then be used with a sorting algorithm
+    // without affecting the sessions in the table itself.
+    //
+    Platform::ScopedMemoryBufferWithSize<SortableSession> sortableSessions;
+    sortableSessions.Calloc(mEntries.Allocated());
```

Comment:
How bug are the entries here?  Would it make sense to instead just have this as a preallocated pool?  I know we don't need it most of the time, but allocating memory when we are doing eviction, and having no way to handle failure to allocate that memory, is making me very nervous.

Reply:
So, my branch original had a pre-allocated pool, and I only switched to heap-based since I was on the fence on the impact to the stack this would have.

Currently as it stands, each `SortableSession` object takes up a single pointer-width (4 bytes on 32-bit platform). On a product that doesn't change the default values in the config, this would take up (5 * 3 + 2) * (4) = 68 bytes.

However, I was planning on adding two more fields in there that are needed for the v1.0 policy algo: a `NumSessionsOnFabric` and `NumNodesOnFabric` stat values that are needed. Those will each take up an additional 2 bytes (16-bits should be enough), meaning we have a total size of (5 * 3 + 2) * (4 + 4) = 132 bytes.

132 bytes on the stack? That felt a little like I was pushing it, hence why I went with heap.

Reply:
If we are getting squeezed, we could distinguish between sessions initiated on this node and those incoming from the listener.  The listener happens from session manager itself, so by definition is lower is app layer.  If we don't have 132 bytes of stack at this point, we are in big trouble.  I would think it is definitely reasonable to allocate 132 bytes on the stack at this point if we are starting from any reasonable position.  Also, won't sigma1 be held in a pbuf or similar?  So I don't think being the recipient of an incoming session establishment request likely puts us in a dangerous stack situation.

Conversely, for outbound sessions, I understand the caution, as these can be initiated from any stack depth, so heap allocation makes sense.  But another nice thing about distinguishing between the two cases is that we really only need to guarantee success for in-bound.

Reply:
Either way, I am OK with doing these changes in a followup, in the interests of getting this stood up for now.